### PR TITLE
Feat/reference external entityRef for techdocs

### PR DESCRIPTION
**Pull request by jrwpatterson**
## Hey, I just made a Pull Request!

Hi there this pr allows the ability to reference other entities techdocs. The reason for this is because we have lots of entities in the same repos that we would like to be able to reference Techdocs from. However if we reference the techdocs from subcomponents we not only have to build and store a new techdoc for each child entity but each of those docs has a link in the techdocs catalog page.

This way we are introducing the `backstage.io/techdocs-external-ref` known annotation. This can be used as an alternative to the `backstage.io/techdocs-ref` to allow it to pickup external docs. This solves two problems firstly not causing the techdocs plugin to try to build the techdocs, also it means the external techdocs don't appear in the techdocs catalog view. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/features/software-catalog/well-known-annotations.md
Comment (freben) [inline docs/features/software-catalog/well-known-annotations.md:103]: This seems to be a copy paste error.

File: docs/features/software-catalog/well-known-annotations.md
Comment (freben) [inline docs/features/software-catalog/well-known-annotations.md:109]: Is the annotation name at risk of being confusing? This "ref" is very unlike the other "ref" above. Other suggestions:

- `backstage.io/documented-by`: make this not about techdocs specifically, but as a general way of pointing elsewhere
- `backstage.io/techdocs-entity`
- keep using `backstage.io/techdocs-ref` with a different target type - like `backstage.io/techdocs-ref: entity:component:/default:foo`
- ...more?

What if were possible to just say `backstage.io/documented-by: system` as a shorthand 🤔 referring to your own system without having to name it fully. Or indeed `backstage.io/techdocs-ref: system:*` or whatever

**Review by freben — COMMENTED** [review]
OK just some thoughts to get started, but leaving to the area maintainers to do the actual review and decisions

File: docs/features/software-catalog/well-known-annotations.md
Comment (jrwpatterson) [reply docs/features/software-catalog/well-known-annotations.md]: Sure I can do that I've gone with the long form to be explicit, as it might not be a system we know developers can be lazy and then you would have to track the relations causing a catalog call when building it. 

I figured it would require being opinionated how people structure their projects, and it might be confusing what if the system is several layers up so you might have System at the top, and this might be a subcomponent of a subcomponent.  This could also be a repo of subcomponents and not in a system. I figured if you are already writing system the specific system wouldn't be too much harder than the risk of picking the wrong system. 

The reason for leaving it outside the existing techdocs-ref, is it allows this change to be small and non confusing. inside the techdocs ref it would mean adding filtering to techdocs pages and a few other parts. This is explicitly in the case if you want the techdocs to be run by another component

I like the suggestion of `techdocs-entity` and have changed so that message as again its explicitly pointing at an entity

File: docs/features/software-catalog/well-known-annotations.md
Comment (jrwpatterson) [reply docs/features/software-catalog/well-known-annotations.md]: whoops I'll fix that now!

**Comment by jrwpatterson** [comment]
Not Stale

**Comment by jrwpatterson** [comment]
@freben what do I need to do now to get this pr approved?

**Comment by agentbellnorm** [comment]
Hey @jrwpatterson, thanks for the PR!

**General Thoughts:**
* It kind of goes against the principle of colocating docs with code. I can see your use case happening though, so I’d rather lean to the pragmatic and let people do this if it is beneficial - it’s not mandatory.
* Impact on Search. I don't think there will be any problems, the catalog entry will still be indexed so the component-name will be searchable, but I suppose no documentation entries will show up for the components with this annotation?

**Needs adressed**
*  Entity page Docs [tab](https://demo.backstage.io/catalog/default/component/backstage/docs). I think this needs to be adressed before merging this. How do you think this tab should behave?

**Question:**
* Do you see any other scenario than components(etc) referring to systems? Like is there a need for a solution this generic? Or could we just update the 404 page with either a redirect or suggest the system documentation? This could potentially benefit more users since they wouldn't have to add the annotations.

**Comment by jrwpatterson** [comment]
> Hey @jrwpatterson, thanks for the PR!
> 
> **General Thoughts:**
> 
> * It kind of goes against the principle of colocating docs with code. I can see your use case happening though, so I’d rather lean to the pragmatic and let people do this if it is beneficial - it’s not mandatory.
> * Impact on Search. I don't think there will be any problems, the catalog entry will still be indexed so the component-name will be searchable, but I suppose no documentation entries will show up for the components with this annotation?
> 
> **Needs adressed**
> 
> * Entity page Docs [tab](https://demo.backstage.io/catalog/default/component/backstage/docs). I think this needs to be adressed before merging this. How do you think this tab should behave?
> 
> **Question:**
> 
> * Do you see any other scenario than components(etc) referring to systems? Like is there a need for a solution this generic? Or could we just update the 404 page with either a redirect or suggest the system documentation? This could potentially benefit more users since they wouldn't have to add the annotations.

@agentbellnorm Thank you for the feedback.

I think in the case of monorepos with multiple systems it means you could end up with a mess of documentation if you want to try to keep documentation in one spot and its confusing to our users where to get the documentation from. Plus from a general aesthetic it looks like a car you haven't optioned with all the missing doc spots 😆 !! It also means when you have a system which has multiple components, eg infra, api, FE. but as they all need to be booted/tested/shipped together there should only be one documentation.

I did think about using parent refs and finding systems. I think I had two concerns
- existing users where the component doesn't need documentation and/or the owning system's document isn't appropriate
- deep dependencies, eg where you have a setup below where we might have to travel long way up the chain to find the parent system

I could suggest a compromise a best-of-both-worlds approach. 

1. If we get this one in for now because it is explicit with the documentation you want, it doesn't affect any existing systems, including search, or the documentation page, and it allows you to build out documentation explicitly. 

2. I then create a discussion off the back of this, whether we take @freben's suggestion you could wild card it with something like adding just `system` if you wanted to use it, your suggestion of either referencing the system if there is an owning system or making the template point at that system. And finally, if we automate it, how to opt-out where it doesn't make sense? Once the discussion is settled and we all agree on a way forward that won't affect existing systems and won't do annoying things like duplications in the techdocs pages, then I'm happy to do the phase 2 for a future release that would implement the changes desired. 

This has been a bit of a bug bear for my colleagues and myself as we have decided where possible to use monorepos on the basis that it is much easier to do branch deployments and ensure changes where the backend and the frontend get released as one get shipped and tested together. 

Screenshot 2023-08-23 at 8 14 46 am

**Comment by jrwpatterson** [comment]
@agentbellnorm  The entity page docs tab has been updated if you put the reference in that also picks up the techdocs-entity annotation

**Review by agentbellnorm — APPROVED** [review]
Ah! My bad, let's go then! We just need @backstage/catalog-maintainers blessing.

**Comment by jrwpatterson** [comment]
@agentbellnorm 🎉 looking forward to it

File: plugins/techdocs/src/EntityPageDocs.tsx
Comment (freben) [inline plugins/techdocs/src/EntityPageDocs.tsx:33]: should probably do `const otherEntity = entity.metadata.annotations?.[TECHDOCS_EXTERNAL_ANNOTATION]` to leverage the constant and only do it once

File: plugins/techdocs/src/Router.tsx
Comment (freben) [inline plugins/techdocs/src/Router.tsx:87]: Just noting that the messaging here will be a bit odd, it'll [tell the user](https://github.com/backstage/backstage/blob/12e081d5c51fb58929b6748f1b38abb01bf2025b/packages/core-components/src/components/EmptyState/MissingAnnotationEmptyState.tsx#L89) `You need to add` both those annotations. Maybe not much to do about that now

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (freben) [inline plugins/catalog/src/components/AboutCard/AboutCard.tsx:117]: should probably do `const otherEntity = entity.metadata.annotations?.[TECHDOCS_EXTERNAL_ANNOTATION]` to leverage the constant and only do it once

**Review by freben — COMMENTED** [review]
Alright just a nit or two, handing over to the techdocs maintainers

File: plugins/techdocs/src/Router.tsx
Comment (jrwpatterson) [reply plugins/techdocs/src/Router.tsx:87]: Yup its something thats difficult maybe I remove it?

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (jrwpatterson) [reply plugins/catalog/src/components/AboutCard/AboutCard.tsx:117]: done

File: plugins/techdocs/src/Router.tsx
Comment (jrwpatterson) [reply plugins/techdocs/src/Router.tsx]: figured remove it and let the doco work and not confuse people as the primary way of doing it will be the Techdocs not the entity

File: plugins/techdocs/src/EntityPageDocs.tsx
Comment (jrwpatterson) [reply plugins/techdocs/src/EntityPageDocs.tsx]: done

**Comment by jrwpatterson** [comment]
thanks @freben given the build passes I think that works?

**Comment by fhp** [comment]
First of all, thanks for creating this PR, this was exactly what I was looking for!

I do have one additional suggestion, that might make this feature even more useful: add the ability to deeplink into a techdoc of another component.

An example / my use-case:
We have a system with several sub-components, that are all described in a single techdocs document. In this document, each sub-component has it's own page (along with some other pages that describe the system as a whole).
In an ideal situation, the system would contain the `techdocs-ref` annotation, and the sub-systems have a `techdocs-entity` reference to the system, and an additional `techdocs-page` (?) annotation that makes it possible to directly link to the page about that sub-component.

I am willing to take a look at this. I am not sure if it's best to add it to this PR, or create a new PR once this one is merged.

**Comment by jrwpatterson** [comment]
> First of all, thanks for creating this PR, this was exactly what I was looking for!
> 
> I do have one additional suggestion, that might make this feature even more useful: add the ability to deeplink into a techdoc of another component.
> 
> An example / my use-case: We have a system with several sub-components, that are all described in a single techdocs document. In this document, each sub-component has it's own page (along with some other pages that describe the system as a whole). In an ideal situation, the system would contain the `techdocs-ref` annotation, and the sub-systems have a `techdocs-entity` reference to the system, and an additional `techdocs-page` (?) annotation that makes it possible to directly link to the page about that sub-component.
> 
> I am willing to take a look at this. I am not sure if it's best to add it to this PR, or create a new PR once this one is merged.

I think that would be a bigger change so I think that would be more appropriate in a seperate PR. I like where you are going though

**Review by freben — APPROVED** [review]

**Comment by jrwpatterson** [comment]
@freben thank you!

@fhp if you want to reach out and work on that pr I'm happy to help I've added my email to my public profile

**Comment by fhp** [comment]
@jrwpatterson Never got around to working on it myself, but someone else did: https://github.com/backstage/backstage/pull/29760

**Pull request merged**
